### PR TITLE
ui/map: update destination on param change

### LINKED ISSUE #28826: nav entry: destination not updated on param change

### PR DESCRIPTION
**Pull request by deanlee**
resolve: https://github.com/commaai/openpilot/issues/28826#issuecomment-1636520930

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
Ready for review?

**Comment by deanlee** [comment]
Yeah. But QFileSystemWatcher::fileChanged will be emitted twice (one for content change, one for last modified time changed) when the param  changes. This causes the list to be cleared and refreshed twice.

File: selfdrive/ui/qt/maps/map_settings.cc [hunk 1]
```diff
@@ -297,6 +298,9 @@ NavigationRequest::NavigationRequest(QObject *parent) : QObject(parent) {
       QObject::connect(repeater, &RequestRepeater::requestDone, this, &NavigationRequest::parseLocationsResponse);
     }
     {
+      auto file_watcher = new QFileSystemWatcher(this);
+      QObject::connect(file_watcher, &QFileSystemWatcher::fileChanged, this, &NavigationRequest::nextDestinationUpdated);
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/maps/map_settings.cc:302]: can we make a custom param watcher so this is easier to reuse?

**Comment by adeebshihadeh** [comment]
> y. But QFileSystemWatcher::fileChanged will be emitted twice (one for content change, one for last modified time changed) when the param changes. This causes the list to be cleared and refreshed twice.

Maybe we can a custom Param watcher class that only emits this once, but I agree that the clearing should be fixed anyway.

Also, maybe we should move the definition of `clearLayout` to the wifi settings so no one else is tempted to use it? then we can get rid of it entirely after that.

> I had planned to open this pr after #28961 was REVIEWED.

@incognitojam can you review and merge both of these tomorrow?

**Comment by adeebshihadeh** [comment]
Thanks @deanlee!

**Comment by adeebshihadeh** [comment]
Still doesn't seem to work completely. I suspect it's not adding back the watcher sometimes after the param is deleted?

**Comment by deanlee** [comment]
`. I suspect it's not adding back the watcher sometimes after the param is deleted?`

y, how about not remove "NavDestination", just set it to empty? 
Otherwise we can only monitor the entire `params/d` directory and call `param_watcher->addParam("NavDestination");` after the NavDestination is recreated. this is very resource consuming.

**Comment by adeebshihadeh** [comment]
We definitely don't want to monitor the whole directory. And I don't like just setting it to empty; it's effectively making a second implementation of Params::remove to enforce behavior elsewhere. 


Thinking about this more, the bug is when the destination is removed, so I don't think we need either of those anyway. Seems like we're just dropping the delete event sometimes?

**Comment by deanlee** [comment]
created a pr to fix this sync issue: https://github.com/commaai/openpilot/pull/29103

**Pull request merged**